transformer_engine/pytorch/tensor/_internal/hybrid_nvfp4_tensor_base.py

```diff
@@ -0,0 +1,219 @@
+# Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+"""Mixin class holding data specific for HybridNVFP4Tensor"""
+
+from __future__ import annotations
+from typing import Optional, Dict, Any, Tuple
+import torch
+
+import transformer_engine_torch as tex
+from transformer_engine_torch import DType as TE_DType
+
+from ..quantized_tensor import QuantizedTensorBase
+
+from ...constants import TE_DType as torch_to_transformer_engine_dtype
+
+from ..quantized_tensor import Quantizer
+
+from ...utils import _empty_tensor
+
+
+class _FromHybridNVFP4Func(torch.autograd.Function):
+    """Cast from NVFP4 to other dtype"""
+
+    @staticmethod
+    def forward(
+        _ctx: Optional[torch.autograd.function.FunctionCtx],  # unused
+        tensor: HybridNVFP4TensorBase,
+        dtype: torch.dtype,
+    ) -> torch.Tensor:
+        # pylint: disable=missing-function-docstring
+        dtype = torch_to_transformer_engine_dtype[dtype]
+
+        # Make sure FP8 data is in expected format
+        if tensor._rowwise_data is not None:
+            return tex.dequantize(tensor, dtype)
+        raise NotImplementedError("Casting back from the transpose not implemented yet!")
+
+    @staticmethod
+    def backward(
+        _ctx: torch.autograd.function.FunctionCtx,  # unused
+        grad: torch.Tensor,
+    ) -> Tuple[Optional[torch.Tensor], ...]:
+        # pylint: disable=missing-function-docstring
+        # Assume that we want gradients in full precision
+        return grad, None
+
+
+class HybridNVFP4TensorBase(QuantizedTensorBase):
+    """Mixin class that holds data attributes of HybridNVFP4Tensor.
+
+    HybridNVFP4Tensor inherits from the PyTorch tensor class and this mixin
+    class. If this class is instantiated directly, it has the same
+    data, lower CPU overhead, and less functionality. It should only
+    be instantiated directly for performance-critical internal usage.
+
+    """
+
+    _rowwise_data: Optional[torch.Tensor]
+    _columnwise_data: Optional[torch.Tensor]
+    _quantizer: Optional[Quantizer]
+    _fp8_dtype: TE_DType
+    _rowwise_scale_inv: torch.Tensor
+    _columnwise_scale_inv: torch.Tensor
+    _per_tensor_rowwise_scale_inv: torch.Tensor
```

Comment:
I would suggest making all the scale-inv names explicit to reduce confusion:
```suggestion
    _rowwise_block_scale_inv: torch.Tensor
    _rowwise_tensor_scale_inv: torch.Tensor
    _columnwise_block_scale_inv: torch.Tensor
```